File: 백준/Gold/5052. 전화번호 목록/전화번호 목록.py
import sys
from collections import defaultdict

T = int(sys.stdin.readline())

for _ in range(T):
    n = int(sys.stdin.readline())
    phone_numbers = []
    for i in range(n):
        phone_numbers.append(sys.stdin.readline().rstrip())

    phone_number_dict = defaultdict(bool)
    phone_numbers.sort(key=lambda x: len(x))
    flag = False
    for phone_number in phone_numbers:
        temp = ''
        for i in range(len(phone_number)):
            temp += phone_number[i]
            if phone_number_dict[temp]:
                flag = True
                break
        if flag:
            break
        phone_number_dict[phone_number] = True

    if flag:
        print('NO')
    else:
        print('YES')

File: 백준/Gold/5052. 전화번호 목록/전화번호 목록.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;

class Main {

    static class Trie {
        char ch;
        boolean isLeaf;
        Trie[] child;

        public Trie() {
            isLeaf = false;
            child = new Trie[10];
        }

        public Trie(char ch) {
            this.ch = ch;
            isLeaf = false;
            child = new Trie[10];
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int T = Integer.parseInt(br.readLine());
        StringBuilder answer = new StringBuilder();
        while (T-- > 0) {
            int n = Integer.parseInt(br.readLine());

            String[] phoneNumbers = new String[n];

            for (int i = 0; i < n; i++) {
                phoneNumbers[i] = br.readLine();
            }

            Arrays.sort(phoneNumbers, (o1, o2) -> o1.length() - o2.length());
            answer.append(solution(phoneNumbers));
        }

        System.out.print(answer);

        br.close();
    }

    static String solution(String[] phoneNumbers) {
        Trie root = new Trie();
        for (String phoneNumber : phoneNumbers) {
            if (trie(root, 0, phoneNumber))
                return "NO\n";
        }
        return "YES\n";
    }

    static boolean trie(Trie parent, int idx, String phoneNumber) {
        if (idx == phoneNumber.length()) {
            parent.isLeaf = true;
            return false;
        }

        if (parent.isLeaf)
            return true;

        char ch = phoneNumber.charAt(idx);

        if(parent.child[ch - '0'] == null){
            parent.child[ch - '0'] = new Trie(ch);
        }

        return trie(parent.child[ch - '0'], idx + 1, phoneNumber);
    }
}